### PR TITLE
m6809: add undocumented opcodes

### PR DESCRIPTION
**Pull request by tlindner**
I used this as a reference: https://github.com/hoglet67/6809Decoder/wiki/Undocumented-6809-Behaviours

The previous pull request #10868 was broken by me after a rebase. I decided to address all of Vas' commented into this new PR.

### CODE REVIEW

**Comment by tlindner** [comment]
I addresses @cuavas  's comments below. All commits done.

> Using greater/less comparisons on “opcode levels”
> doesn’t really make sense when the “levels” aren’t
> simple supersets/subsets and tends to make the code
> confusing.

This was fixed when I switched to using a transparent comparator. The list of opcodes became filtered at the start.

> This change also doesn’t give a good indication in
> the disassembly when there are invalid prefixes, and
> it will behave badly if there‘s a long sequence of
> consecutive prefixes.

I thought of this. I rejected it because the first page code is what is significant.
So your suggestion would produce disassembly like this:
```
0: 10 	           FCB $10
1: 11 83 55 55     CMPD #$5555
5: 11 83 55 55     CMPU #$5555
```
That makes it look like the same sequence of bytes can disassemble to different opcodes.
I like better how the following looks:
```
0: 10 11 83 55 55  CMPD   #$5555
5: 11 83 55 55     CMPU   #$5555
```
I also thought of embedding ignored page opcodes into the mnemonic:
```
0: 10 11 10 11 83 55 55  CMPD-323 #$5555
```
But it didn't seem necessary.

> m_hcf is a bad name for the variable indicating that
> the CPU is in a state where it fetches opcodes and
> does nothing besides increment the program counter.
> It’s completely non-descriptive. My first instinct is
> to read it as “half-carry flag”.

I changed it to m_freerun

> The greater than/less than level comparisons don’t
> really make that much sense any more, as they’re no
> longer supersets in ascending order.

Fixed.

> Doing a level comparison like this means you always
> scan the entire array if there’s no matching opcode.
> Have you considered it might be better to use a set
> with a transparent comparator to avoid the O(n)
> walks?

Fixed.

> This will give no indication in the disassembly that
> the instruction has a technically invalid opcode
> prefix. Maybe you should disassemble the prefix as a
> .db directive or something?

Fixed.

> Also, is it really necessary to misspell “through”?

Comment removed when code was removed.

> This is going to yield a kind of denial-of-service
> attack if there’s a very long sequence of successive
> page 2/page 3 prefixes.

Added check to only disassemble 65535 bytes for each instruction.

**Comment by tlindner** [comment]
I was hoping this could be considered for merging this month. The build failure doesn't seem like a problem with the code. I've compiled for both macOS and Windows.

File: src/devices/cpu/m6809/m6809.lst [hunk 1]
```diff
@@ -600,6 +646,113 @@ TFR:
...
+	{
+		@dummy_read_opcode_arg(0);
+		;
+	}
+	@write_operand(0);
+	return;
+
+XST8:
+	// ToDo: Flag setting is much more complex see:
+	// https://github.com/hoglet67/6809Decoder/wiki/Undocumented-6809-Behaviours#store-immediate
```
Comment (cuavas) [inline src/devices/cpu/m6809/m6809.lst:710]: Please put `TODO` in all caps.  It’s recognised by various text editors and tools.

File: src/devices/cpu/m6809/6x09dasm.h [hunk 2]
```diff
@@ -67,33 +68,46 @@ class m6x09_base_disassembler : public util::disasm_interface
...
 		m6x09_addressing_mode mode() const { return m_mode; }
-		m6x09_instruction_level level() const { return m_level; }
+		uint32_t level() const { return m_level; }
 		unsigned flags() const { return m_flags; }
 		const char *name() const { return m_name; }
 
+		bool operator < (const opcodeinfo & rhs) const
+		{
+			return (opcode() < rhs.opcode());
+		}
```
Comment (cuavas) [inline src/devices/cpu/m6809/6x09dasm.h:94]: You don’t need this when you’ve got a comparator structure.

File: src/devices/cpu/m6809/6x09dasm.h [hunk 3]
```diff
@@ -67,33 +68,46 @@ class m6x09_base_disassembler : public util::disasm_interface
...
-		m6x09_addressing_mode   m_mode : 6;     // Addressing mode
-		m6x09_instruction_level m_level : 2;    // General, or 6309 only?
-		unsigned                m_flags;        // Disassembly flags
-		const char *            m_name;         // Opcode name
+		uint16_t                m_opcode;         // 8-bit opcode value
+		uint8_t                 m_operand_length; // Opcode length in bytes
+		m6x09_addressing_mode   m_mode : 6;       // Addressing mode
+		uint32_t                m_level;          // General, or 6309 only?
+		unsigned                m_flags;          // Disassembly flags
+		const char *            m_name;           // Opcode name
```
Comment (cuavas) [inline src/devices/cpu/m6809/6x09dasm.h:110]: The bitfields were being used in an attempt to encourage the compiler to pack the two fields into a single byte so they’d get packed into a word with the opcode and length.  That won’t happen any more, so there’s no point giving a field width.

You can specify a storage class for `m6x09_addressing_mode` at declaration if you want to ensure it’s stored as a byte.

Also, is there any point having `m_level` take a whole 32 bits?  Do you expect to have that many instruction categories?

File: src/devices/cpu/m6809/6x09dasm.cpp [hunk 4]
```diff
@@ -49,50 +49,107 @@ u32 m6x09_base_disassembler::opcode_alignment() const
...
+				{
+					// multiple pages are illegal on the 6309
+					return nullptr;
+				}
+
+				if (infloop++ > 65534)
+				{
+					// detect infinite loop of page opcodes
+					return nullptr;
+				}
```
Comment (cuavas) [inline src/devices/cpu/m6809/6x09dasm.cpp:142]: As far as I can tell, this still isn’t going to clearly show the presence of the prefixes that shouldn’t be there.  I still think it’s better to disassemble it as a sequence of `.db` directives or something followed by the actual instruction with the final prefix even if the CPU gets into a state where it won’t execute that way.  Folding multiple prefixes like this is going to produce output that’s difficult to use when you try to use it on something that isn’t entirely filled with valid code.

**Comment by gp48k** [comment]
Despite comments I made elsewhere, I think your suggestion of this would be best:
`0: 10 11 10 11 83 55 55  CMPD-323 #$5555
It reflects reality.  Assembling the disassembly will produce the same binary.  And it won't break the debugger "over" command.
However, a single instruction could end up spanning (by accident) most of memory.   Or all of it if there is some pure-RAM 6809 machines out there.  That may present practical problems.  I also don't know if "over" would be broken with other layouts but it at least needs to be checked.

**Comment by npwoods** [comment]
> I like better how the following looks:
>
> ````
> 0: 10 11 83 55 55  CMPD   #$5555
> 5: 11 83 55 55     CMPU   #$5555
> ````

I agree with @tlindner; this is the way to go.  IMHO these extra prefix bytes should just be seen as just yet another way to encode instructions (in the above case, `CMPD`).  Furthermore, the prefix bytes are not true no-ops because interrupts won't be triggered after them; they will only be triggered when the "full" instruction is completed.

Yes there are degenerate cases, such as when we wrap around the end of memory (which I don't think MAME handles well anyways), and the truly bizarre situation where the entire address space is filled with `$10` or `$11`, but it seems silly to let them drive things (though we need to handle them)

File: src/devices/cpu/m6809/m6809.lst (hunk 1 above)
Comment (tlindner) [reply src/devices/cpu/m6809/m6809.lst]: Addressed. Committed.

File: src/devices/cpu/m6809/6x09dasm.h (hunk 2 above)
Comment (tlindner) [reply src/devices/cpu/m6809/6x09dasm.h]: Addressed. Committed.

File: src/devices/cpu/m6809/6x09dasm.h (hunk 3 above)
Comment (tlindner) [reply src/devices/cpu/m6809/6x09dasm.h]: Addressed. Committed.

**Comment by cuavas** [comment]
The debugger’s “next instruction” command isn’t affected by the disassembly – it will just step to the next instruction as interpreted by the CPU (i.e. the next time the CPU calls the debugger instruction hook).

Making up weird mnemonics to represent instructions with invalid prefixes isn’t going to produce the same code from assembling the disassembly unless you get all the assemblers to deal with it as well.  It’s still going to produce impractically long lines in cases where there are large runs of 0x10/0x11.

It’s better to make it obvious to the developer that the bytes are there when they probably shouldn’t be if what you’re disassembling is code.  Just disassemble them as `.db` directives (or whatever the most common directive for inserting arbitrary byte is for 6809 assemblers).

**Comment by cuavas** [comment]
Also, this no longer builds with the changes to `standard_irq_callback` made by @ajrhacker – please update it.

File: src/devices/cpu/m6809/6x09dasm.cpp [hunk 5]
```diff
@@ -49,50 +49,107 @@ u32 m6x09_base_disassembler::opcode_alignment() const
...
+			// on the 6809 an unimplemented page 2 or 3 opcodes fall through to the first page.
+			if (!(m_level & HD6309_EXCLUSIVE) && (page != 0))
+			{
+				// backup the opcode pointer and disassemble the bare opcode.
+				--p;
+				page = 0;
+			}
+			else
+				// nothing to disassemble.
+				return nullptr;
```
[the remark this answers is not in the document]
Comment (tlindner) [reply src/devices/cpu/m6809/6x09dasm.cpp]: Implemented @cuavas 's suggestion.

File: src/devices/cpu/m6809/6x09dasm.cpp (hunk 4 above)
Comment (tlindner) [reply src/devices/cpu/m6809/6x09dasm.cpp]: Implemented @cuavas 's suggestion

**Comment by tlindner** [comment]
Committed patch to address `standard_irq_callback`. I had to `git pull upstream master` so it brought along all the recent changes. Hopefully I have not ruined this PR.

**Comment by tlindner** [comment]
Are the conflicts listed here really a problem? I have not edited any of these files.

**Pull request closed without merging**